### ISSUE TITLE
ElasticBeanstalk.describeEnvironments does not provide environment resource details.

### ISSUE THREAD

**Issue by kunagpal** [aws/aws-sdk-js#1073]
> **Issue metadata**
**Node** `v4.3.0`
**AWS-SDK** `v2.4.9`
**OS** `Ubuntu 16.04 64 bit`

As per the official AWS-SDK for nodejs documentation provided [here](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ElasticBeanstalk.html#describeEnvironments-property), `ElasticBeanstalk.describeEnvironments` provides an object with the key `Environments` as an `Array`, with each element having a key `Resources`.

However, a few recent calls made to this method have resulted in objects where elements of the `Environments` array do not contain the key `Resources`. A sample snippet used for the purpose described above is as follows:

```javascript
var AWS = require('aws-sdk');
 // access tokens picked up from the environment or ~/.aws/credentials
AWS.config.update({region: 'region'});
var ElasticBeanstalk = new AWS.ElasticBeanstalk();

ElasticBeanstalk.describeEnvironments(function (err, data) {
        // handle errors
        console.log(data.Environments);
});
```

The contents logged to the console for the snippet above are as follows:
```javascript
[
    {
        EnvironmentName: 'envName',
        EnvironmentId: 'id',
        ApplicationName: 'appName',
        VersionLabel: 'label',
        SolutionStackName: 'solStack',
        Description: 'description...',
        EndpointURL: 'URL',
        CNAME: 'CNAME',
        DateCreated: TimeStamp,
        DateUpdated: TimeStamp,
        Status: 'Ready',
        AbortableOperationInProgress: false,
        Health: 'Green',
        HealthStatus: 'Ok',
        Tier: { Name: 'WebServer', Type: 'Standard', Version: '' },
        EnvironmentLinks: []
    }
]
```

In other words, the `Resources` key is missing. Is this an issue with the implementation, or with the SDK code / documentation?

**Comment by LiuJoyceC** [comment]
Hi @kunagpal 
Thanks for your question. The API docs give descriptions of each property you might get back from the response, but it is not a list of guaranteed properties. The SDK does not require the service to send back certain properties in their response, so you should always check for the existence of a property before using it. This applies to all services, not just Elastic Beanstalk. I hope that answers your question! Please let me know if you have more questions.

**Comment by kunagpal** [comment]
@LiuJoyceC
Thanks for the explanation, this clears a lot of issues. Out of curiosity, under what circumstances would the aforementioned method call actually provide the mentioned `Resources`?

**Comment by LiuJoyceC** [comment]
Hi @kunagpal 

You mentioned your recent calls did not return the `Resources` key. On what date did you notice this change? And before that, were you consistently getting back the `Resources` key for the same calls? I'd like to investigate if there is a change in behavior on the service side.

If you need the information on the environment resources, you can call `describeEnvironmentResources` and pass in the name or ID for the environment.

**Comment by kunagpal** [comment]
@LiuJoyceC: The absence of the `Resources` field has been consistent ever since I've started using `ElasticBeanstalk.describeEnvironments`. Since the `Resources` key was missing from square one, I did switch over to `ElasticBeanstalk.describeEnvironmentResources`. However, the nature of `ElasticBeanstalk.describeEnvironmentResources` requires that an environment id or name be specified, for which `describeEnvironments` needs to be called first. In such a case, having `Resources` served within `describeEnvironments` would have save one extra AWS-SDK API call.

**Comment by LiuJoyceC** [comment]
Hi @kunagpal 

You're right, it would indeed be more convenient to not need to make an extra API call. However, it's possible that adding an extra potentially large object to the response could significantly increase the size of the response and force some customers to have to use pagination (which also requires extra API calls) in cases where they otherwise wouldn't need to when they don't need the `Resources` information. Feel free to post a feature request to the Elastic Beanstalk service team in the AWS Forums: https://forums.aws.amazon.com/forum.jspa?forumID=86, and the service team can determine what solution will best serve most customers' use cases.

Hope that helps. Let me know if there are other SDK-specific questions you have related to this issue.

**Comment by kunagpal** [comment]
@LiuJoyceC: Thanks for all the help, this clears things up quite nicely :smile:
